src/microbe_stage/MicrobeAI.cs

```diff
@@ -370,43 +397,85 @@ private void RunAndTumble(Random random)
         microbe.State = Microbe.MicrobeState.Normal;
 
         var usefulCompounds = microbe.TotalAbsorbedCompounds.Where(x => microbe.Compounds.IsUseful(x.Key));
+        ComputeCompoundsSearchWeights(usefulCompounds.ToList());
```

Comment:
It would be much better to make this `.ToList()` one line above, or, even better in my opinion, use `.ToDictionary(i => i.Key, i => i.Value)` there.

Reply:
I get for the call to `.ToList()` above (I only use it right below anyways), but I don't get the benefit of having a dictionary here, as I never use `usefulCompounds` as such.

If this is for futur usages which may use a dictionary, why not keeping it as a generic IEnumerable, as it could be used as a list as well?

Reply:
I saw that this is a collection of a key value pair, which looks odd to me becasue there is a special datatype in c# for that and `TotalAbsorbedCompounds` is a Dictionary anyways.
`Dictionary<Compound, float>` looks much cleaner to me than `IEnumerable<KeyValuePair<Compound,float>>`.

But looking though the code, it seems that an arraylist should be more more efficient than a hashmap in this case, so you can keep it as a list.

Reply:
So the `IEnumerable<KeyValuePair<Compound,float>>` was the original type of `usefulCompounds` and I did not change it. But given that it's only used as a list at some point I just went ahead and directly converted to List in the var instanciation.

Reply:
Useful compounds purposefully uses a general interface, as it doesn't have a real dictionary to return (at least I think so) because it gets the ammonia etc. added there, so that's why the return type is as is.
ToList should *only* be used if the list is iterated multiple times, otherwise it is a waste.

Reply:
Okay so I had a doubt and tried again, but in fact I tried keeping IEnumerable, but jetbrains did not let me do so. The up to date version (commit 540b708) is using IEnumerables and I got this from auto check:
src\microbe_stage\MicrobeAI.cs:458 Possible multiple enumeration type: PossibleMultipleEnumeration

(for all lines where usefulCompounds is used)

Reply:
If there *are* multiple enumerations, *then* use ToList, if it is enumerated just once (and for example the method call makes the checker confused) then add a suppression.

Reply:
Yeah, I think the function call is what makes the checker confused. I went ahead and used another solution: I actually moved the computation of useful compounds to the search weights computation itself as only it was using this compounds list.